### ISSUE TITLE
Rename precondition.assert* to require*

### ISSUE THREAD

**Issue by mitschabaude** [o1-labs/o1js#1247]
**Why**: The shared "assert" naming between constraints like `x.assertEquals(0)` and preconditions like `this.state.assertEquals(Field(0))` is a constant source of confusion. We settled on using `require()` instead of `assert()` for preconditions to clearly mark it as a distinct concept.

**How**: Change all occurences of assert* in precondition methods, e.g.

* `this.state.assertEquals()` -> `this.state.requireEquals()`
* `this.state.getAndAssertEquals()` -> `this.state.getAndRequireEquals()` (is that too verbose?)
* `this.account.balance.assertBetween()` -> `this.account.balance.requireBetween()`

**Comment by Saurabhpatil-dev** [comment]
Hi @mitschabaude ,

I am get that you want occurrences of assert* to be changed and I would like to contribute.
So, if possible can you please help me understand in which files do you want occurrences changed I will try to understand them and start with it. More of like I want a starting point.

Thank you 😃

**Comment by mitschabaude** [comment]
Hey @Saurabhpatil-dev that would be very welcome!!

The API that we want to change is implemented in two files:
* `src/lib/state.ts` - onchain state
* `src/lib/precondition.ts` - other preconditions

It should be enough to change those two files, and then adapt all tests.

**Important**: Don't delete the old APIs, only deprecate them using the [`@deprecated` doccomment](https://jsdoc.app/tags-deprecated)

**Comment by Saurabhpatil-dev** [comment]
Awesome!!!

Let me try to get it done and up for review by tomorrow.😃

Thanks

**Comment by Saurabhpatil-dev** [comment]
Hi @mitschabaude ,

I have made changes in `src/lib/state.ts` file. #1263 
Can you please help to review them ?

Also, I am not sure about the changes in `src/lib/precondition.ts` file. Can you please help me understand it little more ?

Thank You 😃

**Comment by mitschabaude** [comment]
@Saurabhpatil-dev sure, if you look at `precondition.ts`, you'll find the `assert*` API in a few places, and the goal would be to update all of them. This is what users will see when they do something like `this.account.balance.getAndAssertEquals()` in a smart contract.

Places to change:
* https://github.com/o1-labs/o1js/blob/3261e8c93a44167120b515262f56490767099659/src/lib/precondition.ts#L453-L461
  * I think this is where we want the `@deprecated` tags! it defines the interface
* https://github.com/o1-labs/o1js/blob/3261e8c93a44167120b515262f56490767099659/src/lib/precondition.ts#L52-L71
* https://github.com/o1-labs/o1js/blob/3261e8c93a44167120b515262f56490767099659/src/lib/precondition.ts#L121
* https://github.com/o1-labs/o1js/blob/3261e8c93a44167120b515262f56490767099659/src/lib/precondition.ts#L235-L259

You can see your changes in TS by re-running `npm run build`, and then looking at `src/lib/precondition.test.ts`. You can run that test with `./jest src/lib/precondition.test.ts`

**Comment by Saurabhpatil-dev** [comment]
Hi @mitschabaude , 

Okay, Will make a PR for this once the current PR is merged if it's fine ?

**Comment by Saurabhpatil-dev** [comment]
Hi @mitschabaude ,

While re-building the project locally I got the errors related to this PR you made yesterday #1260 .

![Screenshot 2023-11-22 at 8 14 47 PM](https://github.com/o1-labs/o1js/assets/55076843/4c08f4f3-83c3-499f-a15f-cbbbfb1f5126)


Locally I am just reverting the changes locally for testing. Will not commit it.

**Comment by mitschabaude** [comment]
@Saurabhpatil-dev you need to update the submodule, `git submodule update --init`

**Comment by Saurabhpatil-dev** [comment]
Hey @mitschabaude ,

Thanks it worked.

The  Precondition API's are also renamed and the testing is also completed.
Adding Screenshots for your reference.

Please let know if I should commit changes now or wait for the PR to get merged and then create a new PR ?
Also, do you want me to commit the test file too ?

**Screenshots**
![Screenshot 2023-11-22 at 9 24 54 PM](https://github.com/o1-labs/o1js/assets/55076843/e8ae63d5-ddeb-4fcb-b2fc-d7e695ac6cde)

![Screenshot 2023-11-22 at 9 17 22 PM](https://github.com/o1-labs/o1js/assets/55076843/8f96683c-fb6b-44f9-a82e-c33f037ece67)

![Screenshot 2023-11-22 at 9 31 28 PM](https://github.com/o1-labs/o1js/assets/55076843/d67acf66-e6df-40c4-9818-1acc45c816a0)

**Comment by mitschabaude** [comment]
@Saurabhpatil-dev I'd prefer a new PR. And yes please also commit the test changes. Ideally your next PR would also update most examples in `src/examples` that currently use assert*

**Comment by Saurabhpatil-dev** [comment]
Hi @mitschabaude ,

Created a new PR : #1265

Once this PR get approved and merged I would like to start updating the examples in `src/examples` .
For it will we need a new issue and can I update the examples 😅?? 

Thank You😃

**Comment by Saurabhpatil-dev** [comment]
HI @mitschabaude ,

As discussed I am starting with changing the example files locally. 
I am just going through code and then replacing the deprecated ones with the new. (Not retaining the deprecated one)

Can you please check and let me know if this what you were expecting ?

Here is screenshot:
<img width="1045" alt="Screenshot 2023-11-23 at 10 26 44 PM" src="https://github.com/o1-labs/o1js/assets/55076843/c9989112-d96d-4b80-aead-8e9e88db2db8">

Thank You 😃

**Comment by mitschabaude** [comment]
Yes @Saurabhpatil-dev, exactly what I was expecting 💯

**Comment by Saurabhpatil-dev** [comment]
Hi @mitschabaude,

Okay. Will start committing them. 
Is there anyway I can test them or it's fine?

Thanks 😃

**Comment by mitschabaude** [comment]
They're tested in CI. You can test then locally with ./jest file

**Comment by Saurabhpatil-dev** [comment]
Hi @mitschabaude ,

Testing of the example is in progress.
Just wanted you to check if we need to update the error messages too.
If you will see below we are getting the assertEquals() rather than requireEquals():
```
Transaction failed with error Field.assertEquals(): 
```

1. Please let me know if we need to update them and where can I do that ?
2. Should I commit all the example files in new PR ?

Screen Shot for reference:
<img width="1409" alt="Screenshot 2023-11-24 at 6 48 13 PM" src="https://github.com/o1-labs/o1js/assets/55076843/4849a985-2d7f-48bd-a33f-3f9ac220422e">

**Comment by mitschabaude** [comment]
2. If you'd commit all examples in a new PR, that would be great!

**Comment by LuffySama-Dev** [comment]
Hi @mitschabaude ,

As discussed new PR is created #1269.
Can you please help to review and approve ?

Thank You 😃

**Comment by mitschabaude** [comment]
I think we can call this done! Next step for us is to update zkapp-cli examples and docs

**Comment by LuffySama-Dev** [comment]
Hi @mitschabaude ,

Okay, can you please assign the issue to me once it is raised ?? I would love to start working on it  😃.
Thank you so much for guiding me through out.

Thank You 🙌
